src/bullet.py

```diff
@@ -1,48 +1,87 @@
 
-# This allows a circular import only for the purposes of type hints
-# Weapon will never create and instance of Player
-if TYPE_CHECKING:
-    from weapon import Weapon
-
 
 class Projectile:
-    def __init__(self, sprite: LinkedSprite[Projectile], sprite_lists: SpriteLists):
-        self.sprite = sprite
-        sprite.owner = self
-        sprite_lists.projectiles.append(self.sprite)
-        self.damage = 0
+    """
+    A projectile can be anything that is created by a weapon or another projectile included beams, rockets, etc.
+    It stores the characteristics of the projectile and is responible the associated sprite
+    """
 
+    sprite: LinkedSprite[Projectile]
+    sprite_lists: SpriteLists
+    damage: float
+    speed: float
+    radians: float
+    sprite_rotation_offset: float
+    exists: bool
+    beam: bool
+    beam_range: float
+    explodes: bool
 
-class Beam:
     def __init__(
-        self, sprite: LinkedSprite[Beam], sprite_lists: SpriteLists, weapon: Weapon
+        self,
+        sprite: LinkedSprite[Projectile],
+        sprite_lists: SpriteLists,
+        speed: float,
+        radians: float,
+        damage: float,
     ):
         self.sprite = sprite
         sprite.owner = self
-        sprite_lists.beams.append(self.sprite)
-        self.dps = 0
-        self.weapon = weapon
+        self.sprite_lists = sprite_lists
+        self.speed = speed
+        self.radians = radians
+        self.damage = damage
+        self.beam = False
+        self.beam_range = 0
+        self.explodes = False
+        self.sprite_rotation_offset = 0
+        self.append_sprite()
+
+    def update(self, delta_time: float):
+        move_sprite_polar(self.sprite, self.speed * delta_time, self.radians)
+
+    def set_location(self, location: Tuple[float, float, float]):
+        set_sprite_location(self.sprite, location)
+        self.sprite.radians += self.sprite_rotation_offset
 
+    def set_beam(self, range: float):
+        self.beam = True
+        self.beam_range = range
 
-# For naming purposes, a bullet can be anything that comes out of a weapon included beams, rockets, etc
-def bullet_behavior(
+    def set_sprite_rotation_offset(self, radians: float):
+        self.sprite_rotation_offset = radians
+
+    def set_explodes(self):
+        self.explodes = True
+    
+    def append_sprite(self):
+        self.sprite_lists.projectiles.append(self.sprite)
+        self.exists = True
+
+    def remove_sprite(self):
+        self.sprite_lists.projectiles.remove(self.sprite)
+        self.exists = False
+
+def projectile_behavior(
     delta_time: float,
     sprite_lists: SpriteLists,
 ):
     for projectile_sprite in sprite_lists.projectiles:
         projectile_sprite: LinkedSprite[Projectile]
+        # If beam, skip the current projectile collision code that deletes projectiles that hit walls
+        if projectile_sprite.owner.beam:
+            continue
+        projectile_sprite.owner.update(delta_time)
```

Comment:
Why not call `update` on all projectiles all the time?  Since beams might have update logic in the future.

I guess what I'm saying is: it's a bit messy to have destruction logic here in `projectile_behavior`, and updating logic in `Projectile`.  Maybe this comment is hypocritical given our past conversations.  But I think still worth mentioning.

Reply:
Agreed, This is all being changed in another as the same time, so this is a stop gap to get the code to work without duplicating that effort. If this branch gets merged after that, I'll make this messy logic go away in the merge.

Reply:
Fixed